Make Gamepad.vibrationActuator work with the GameController framework

cdumez opened the pull request:
#### c7f1532632f76bad38f73f3718c79d1363ee7abf
<pre>
Make Gamepad.vibrationActuator work with the GameController framework
<a href="https://bugs.webkit.org/show_bug.cgi?id=250217">https://bugs.webkit.org/show_bug.cgi?id=250217</a>
rdar://problem/103973520

Reviewed by Brady Eidson.

Make Gamepad.vibrationActuator work with the GameController framework, so that
the gamepad now actually vibrates on macOS 13+.

This was tested manually using an XBox controller.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/gamepad/cocoa/CoreHapticsSoftLink.h: Added.
* Source/WebCore/platform/gamepad/cocoa/CoreHapticsSoftLink.mm: Added.
* Source/WebCore/platform/gamepad/cocoa/GameControllerGamepad.h:
* Source/WebCore/platform/gamepad/cocoa/GameControllerGamepad.mm:
(WebCore::GameControllerGamepad::ensureHapticEngines):
(WebCore::GameControllerGamepad::playEffect):
(WebCore::GameControllerGamepad::stopEffects):
(WebCore::GameControllerGamepad::noLongerHasAnyClient):
* Source/WebCore/platform/gamepad/cocoa/GameControllerGamepadProvider.mm:
(WebCore::GameControllerGamepadProvider::stopMonitoringGamepads):
* Source/WebCore/platform/gamepad/cocoa/GameControllerHapticEffect.h: Copied from Source/WebCore/platform/gamepad/cocoa/GameControllerGamepad.h.
* Source/WebCore/platform/gamepad/cocoa/GameControllerHapticEffect.mm: Added.
(WebCore::GameControllerHapticEffect::create):
(WebCore::GameControllerHapticEffect::GameControllerHapticEffect):
(WebCore::GameControllerHapticEffect::~GameControllerHapticEffect):
(WebCore::GameControllerHapticEffect::start):
(WebCore::GameControllerHapticEffect::stop):
(WebCore::GameControllerHapticEffect::strongEffectFinishedPlaying):
(WebCore::GameControllerHapticEffect::weakEffectFinishedPlaying):
* Source/WebCore/platform/gamepad/cocoa/GameControllerHapticEngines.h: Copied from Source/WebCore/platform/gamepad/cocoa/GameControllerGamepad.h.
(WebCore::GameControllerHapticEngines::strongEngine):
(WebCore::GameControllerHapticEngines::weakEngine):
* Source/WebCore/platform/gamepad/cocoa/GameControllerHapticEngines.mm: Added.
(WebCore::GameControllerHapticEngines::create):
(WebCore::GameControllerHapticEngines::GameControllerHapticEngines):
(WebCore::GameControllerHapticEngines::playEffect):
(WebCore::GameControllerHapticEngines::stopEffects):
(WebCore::GameControllerHapticEngines::ensureStarted):
(WebCore::GameControllerHapticEngines::stop):

Canonical link: <a href="https://commits.webkit.org/258674@main">https://commits.webkit.org/258674@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2aac1aaef657945ecff1414ebf4e5b102608903d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/102650 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/11780 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/35685 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/111917 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/172140 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/106618 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/12785 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/2686 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/94925 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/109618 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/108426 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/9811 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/93024 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/37473 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/91663 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/24535 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/79216 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/5236 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/25954 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/5396 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/2406 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/11417 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/45443 "Passed tests") | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/74/builds/5958 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/7126 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3165 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

beidson commented on Source/WebCore/platform/gamepad/cocoa/GameControllerHapticEngines.mm:
```diff
@@ -0,0 +1,137 @@
+}
+
+GameControllerHapticEngines::GameControllerHapticEngines(GCController *gamepad)
+    : m_strongEngine([gamepad.haptics createEngineWithLocality:get_GameController_GCHapticsLocalityLeftHandle()])
+    , m_weakEngine([gamepad.haptics createEngineWithLocality:get_GameController_GCHapticsLocalityRightHandle()])
```
I don't get the naming here - Strong == left, Weak == right?

cdumez replied:
Yes, this is what was recommended to me by one of our GameController engineers.

beidson commented on Source/WebCore/platform/gamepad/cocoa/GameControllerHapticEngines.mm:
```diff
@@ -0,0 +1,137 @@
+            });
+            return CHHapticEngineFinishedActionLeaveEngineRunning;
+        }).get()];
+    }).get()];
```
The duplicated code here is a shame.

cdumez committed:
Make Gamepad.vibrationActuator work with the GameController framework
https://bugs.webkit.org/show_bug.cgi?id=250217
rdar://problem/103973520

Reviewed by Brady Eidson.

Make Gamepad.vibrationActuator work with the GameController framework, so that
the gamepad now actually vibrates on macOS 13+.

This was tested manually using an XBox controller.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/gamepad/cocoa/CoreHapticsSoftLink.h: Added.
* Source/WebCore/platform/gamepad/cocoa/CoreHapticsSoftLink.mm: Added.
* Source/WebCore/platform/gamepad/cocoa/GameControllerGamepad.h:
* Source/WebCore/platform/gamepad/cocoa/GameControllerGamepad.mm:
(WebCore::GameControllerGamepad::ensureHapticEngines):
(WebCore::GameControllerGamepad::playEffect):
(WebCore::GameControllerGamepad::stopEffects):
(WebCore::GameControllerGamepad::noLongerHasAnyClient):
* Source/WebCore/platform/gamepad/cocoa/GameControllerGamepadProvider.mm:
(WebCore::GameControllerGamepadProvider::stopMonitoringGamepads):
* Source/WebCore/platform/gamepad/cocoa/GameControllerHapticEffect.h: Copied from Source/WebCore/platform/gamepad/cocoa/GameControllerGamepad.h.
* Source/WebCore/platform/gamepad/cocoa/GameControllerHapticEffect.mm: Added.
(WebCore::GameControllerHapticEffect::create):
(WebCore::GameControllerHapticEffect::GameControllerHapticEffect):
(WebCore::GameControllerHapticEffect::~GameControllerHapticEffect):
(WebCore::GameControllerHapticEffect::start):
(WebCore::GameControllerHapticEffect::stop):
(WebCore::GameControllerHapticEffect::strongEffectFinishedPlaying):
(WebCore::GameControllerHapticEffect::weakEffectFinishedPlaying):
* Source/WebCore/platform/gamepad/cocoa/GameControllerHapticEngines.h: Copied from Source/WebCore/platform/gamepad/cocoa/GameControllerGamepad.h.
(WebCore::GameControllerHapticEngines::strongEngine):
(WebCore::GameControllerHapticEngines::weakEngine):
* Source/WebCore/platform/gamepad/cocoa/GameControllerHapticEngines.mm: Added.
(WebCore::GameControllerHapticEngines::create):
(WebCore::GameControllerHapticEngines::GameControllerHapticEngines):
(WebCore::GameControllerHapticEngines::playEffect):
(WebCore::GameControllerHapticEngines::stopEffects):
(WebCore::GameControllerHapticEngines::ensureStarted):
(WebCore::GameControllerHapticEngines::stop):

Canonical link: https://commits.webkit.org/258674@main
